ARDUINO/LaCOOLBoard/extras/doxygenDocs/man/man3/CoolPubSubClient.3
.TH "CoolPubSubClient" 3 "Tue Sep 19 2017" "CoolBoardAPI" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CoolPubSubClient
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <CoolPubSubClient\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCoolPubSubClient\fP ()"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (IPAddress, uint16_t, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (IPAddress, uint16_t, Client &client, Stream &)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (IPAddress, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (IPAddress, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client, Stream &)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (uint8_t *, uint16_t, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (uint8_t *, uint16_t, Client &client, Stream &)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (uint8_t *, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (uint8_t *, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client, Stream &)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (const char *, uint16_t, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (const char *, uint16_t, Client &client, Stream &)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (const char *, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP (const char *, uint16_t, \fBMQTT_CALLBACK_SIGNATURE\fP, Client &client, Stream &)"
.br
.ti -1c
.RI "\fB~CoolPubSubClient\fP ()"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetServer\fP (IPAddress \fBip\fP, uint16_t \fBport\fP)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetServer\fP (uint8_t *\fBip\fP, uint16_t \fBport\fP)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetServer\fP (const char *\fBdomain\fP, uint16_t \fBport\fP)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetCallback\fP (\fBMQTT_CALLBACK_SIGNATURE\fP)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetClient\fP (Client &client)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetStream\fP (Stream &\fBstream\fP)"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP & \fBsetTimeout\fP (uint16_t \fBsocketTimeout\fP)"
.br
.ti -1c
.RI "boolean \fBconnect\fP (const char *id, uint16_t \fBkeepAlive\fP=\fBMQTT_KEEPALIVE\fP, uint16_t \fBsocketTimeout\fP=\fBMQTT_SOCKET_TIMEOUT\fP)"
.br
.ti -1c
.RI "boolean \fBconnect\fP (const char *id, const char *user, const char *pass, uint16_t \fBkeepAlive\fP=\fBMQTT_KEEPALIVE\fP, uint16_t \fBsocketTimeout\fP=\fBMQTT_SOCKET_TIMEOUT\fP)"
.br
.ti -1c
.RI "boolean \fBconnect\fP (const char *id, const char *willTopic, uint8_t willQos, boolean willRetain, const char *willMessage, uint16_t \fBkeepAlive\fP=\fBMQTT_KEEPALIVE\fP, uint16_t \fBsocketTimeout\fP=\fBMQTT_SOCKET_TIMEOUT\fP)"
.br
.ti -1c
.RI "boolean \fBconnect\fP (const char *id, const char *user, const char *pass, const char *willTopic, uint8_t willQos, boolean willRetain, const char *willMessage, uint16_t \fBkeepAlive\fP=\fBMQTT_KEEPALIVE\fP, uint16_t \fBsocketTimeout\fP=\fBMQTT_SOCKET_TIMEOUT\fP)"
.br
.ti -1c
.RI "void \fBdisconnect\fP ()"
.br
.ti -1c
.RI "boolean \fBpublish\fP (const char *topic, const char *payload)"
.br
.ti -1c
.RI "boolean \fBpublish\fP (const char *topic, const char *payload, boolean retained)"
.br
.ti -1c
.RI "boolean \fBpublish\fP (const char *topic, const uint8_t *payload, unsigned int plength)"
.br
.ti -1c
.RI "boolean \fBpublish\fP (const char *topic, const uint8_t *payload, unsigned int plength, boolean retained)"
.br
.ti -1c
.RI "boolean \fBpublish_P\fP (const char *topic, const uint8_t *payload, unsigned int plength, boolean retained)"
.br
.ti -1c
.RI "boolean \fBsubscribe\fP (const char *topic)"
.br
.ti -1c
.RI "boolean \fBsubscribe\fP (const char *topic, uint8_t qos)"
.br
.ti -1c
.RI "boolean \fBunsubscribe\fP (const char *topic)"
.br
.ti -1c
.RI "boolean \fBloop\fP ()"
.br
.ti -1c
.RI "boolean \fBconnected\fP ()"
.br
.ti -1c
.RI "int \fBstate\fP ()"
.br
.ti -1c
.RI "boolean \fBsetBufferSize\fP (uint16_t size)"
.br
.ti -1c
.RI "uint16_t \fBgetBufferSize\fP ()"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "uint16_t \fBreadPacket\fP (uint8_t *)"
.br
.ti -1c
.RI "boolean \fBreadByte\fP (uint8_t *result)"
.br
.ti -1c
.RI "boolean \fBreadByte\fP (uint8_t *result, uint16_t *index)"
.br
.ti -1c
.RI "boolean \fBwrite\fP (uint8_t header, uint8_t *buf, uint16_t length)"
.br
.ti -1c
.RI "uint16_t \fBwriteString\fP (const char *string, uint8_t *buf, uint16_t pos)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "Client * \fB_client\fP"
.br
.ti -1c
.RI "uint8_t * \fBbuffer\fP"
.br
.ti -1c
.RI "uint16_t \fBbuffer_size\fP"
.br
.ti -1c
.RI "uint16_t \fBnextMsgId\fP"
.br
.ti -1c
.RI "unsigned long \fBlastOutActivity\fP"
.br
.ti -1c
.RI "unsigned long \fBlastInActivity\fP"
.br
.ti -1c
.RI "bool \fBpingOutstanding\fP"
.br
.ti -1c
.RI "\fBMQTT_CALLBACK_SIGNATURE\fP"
.br
.ti -1c
.RI "IPAddress \fBip\fP"
.br
.ti -1c
.RI "const char * \fBdomain\fP"
.br
.ti -1c
.RI "uint16_t \fBport\fP"
.br
.ti -1c
.RI "Stream * \fBstream\fP"
.br
.ti -1c
.RI "int \fB_state\fP"
.br
.ti -1c
.RI "uint16_t \fBkeepAlive\fP"
.br
.ti -1c
.RI "uint16_t \fBsocketTimeout\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 88 of file CoolPubSubClient\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "CoolPubSubClient::CoolPubSubClient ()"

.PP
Definition at line 14 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (Client & client)"

.PP
Definition at line 23 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (IPAddress addr, uint16_t port, Client & client)"

.PP
Definition at line 31 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (IPAddress addr, uint16_t port, Client & client, Stream & stream)"

.PP
Definition at line 39 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (IPAddress addr, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client)"

.PP
Definition at line 47 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (IPAddress addr, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client, Stream & stream)"

.PP
Definition at line 56 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (uint8_t * ip, uint16_t port, Client & client)"

.PP
Definition at line 66 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (uint8_t * ip, uint16_t port, Client & client, Stream & stream)"

.PP
Definition at line 74 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (uint8_t * ip, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client)"

.PP
Definition at line 82 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (uint8_t * ip, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client, Stream & stream)"

.PP
Definition at line 91 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (const char * domain, uint16_t port, Client & client)"

.PP
Definition at line 101 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (const char * domain, uint16_t port, Client & client, Stream & stream)"

.PP
Definition at line 109 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (const char * domain, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client)"

.PP
Definition at line 117 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::CoolPubSubClient (const char * domain, uint16_t port, \fBMQTT_CALLBACK_SIGNATURE\fP, Client & client, Stream & stream)"

.PP
Definition at line 126 of file CoolPubSubClient\&.cpp\&.
.SS "CoolPubSubClient::~CoolPubSubClient ()"

.PP
Definition at line 136 of file CoolPubSubClient\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "boolean CoolPubSubClient::connect (const char * id, uint16_t keepAlive = \fC\fBMQTT_KEEPALIVE\fP\fP, uint16_t socketTimeout = \fC\fBMQTT_SOCKET_TIMEOUT\fP\fP)"

.PP
Definition at line 140 of file CoolPubSubClient\&.cpp\&.
.SS "boolean CoolPubSubClient::connect (const char * id, const char * user, const char * pass, uint16_t keepAlive = \fC\fBMQTT_KEEPALIVE\fP\fP, uint16_t socketTimeout = \fC\fBMQTT_SOCKET_TIMEOUT\fP\fP)"

.PP
Definition at line 144 of file CoolPubSubClient\&.cpp\&.
.SS "boolean CoolPubSubClient::connect (const char * id, const char * willTopic, uint8_t willQos, boolean willRetain, const char * willMessage, uint16_t keepAlive = \fC\fBMQTT_KEEPALIVE\fP\fP, uint16_t socketTimeout = \fC\fBMQTT_SOCKET_TIMEOUT\fP\fP)"

.PP
Definition at line 148 of file CoolPubSubClient\&.cpp\&.

ARDUINO/LaCOOLBoard/src/internals/CoolPubSubClient.cpp
/*
 CoolPubSubClient.h 
 Modified by Mehdi Zemzem
 -added dynamic buffer and parameters settings

  PubSubClient.cpp - A simple client for MQTT.
  Nick O'Leary
  http://knolleary.net
*/

#include "CoolPubSubClient.h"
#include "Arduino.h"

CoolPubSubClient::CoolPubSubClient() {
    this->_state = MQTT_DISCONNECTED;
    this->_client = NULL;
    this->stream = NULL;
    setCallback(NULL);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}

CoolPubSubClient::CoolPubSubClient(Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}

CoolPubSubClient::CoolPubSubClient(IPAddress addr, uint16_t port, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(addr, port);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(IPAddress addr, uint16_t port, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(addr,port);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(IPAddress addr, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(addr, port);
    setCallback(callback);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(IPAddress addr, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(addr,port);
    setCallback(callback);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}

CoolPubSubClient::CoolPubSubClient(uint8_t *ip, uint16_t port, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(ip, port);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(uint8_t *ip, uint16_t port, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(ip,port);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(uint8_t *ip, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(ip, port);
    setCallback(callback);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(uint8_t *ip, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(ip,port);
    setCallback(callback);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}

CoolPubSubClient::CoolPubSubClient(const char* domain, uint16_t port, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(domain,port);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(const char* domain, uint16_t port, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(domain,port);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(const char* domain, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client) {
    this->_state = MQTT_DISCONNECTED;
    setServer(domain,port);
    setCallback(callback);
    setClient(client);
    this->stream = NULL;
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}
CoolPubSubClient::CoolPubSubClient(const char* domain, uint16_t port, MQTT_CALLBACK_SIGNATURE, Client& client, Stream& stream) {
    this->_state = MQTT_DISCONNECTED;
    setServer(domain,port);
    setCallback(callback);
    setClient(client);
    setStream(stream);
    this->buffer_size = MQTT_MAX_PACKET_SIZE;
    this->buffer = (uint8_t*)malloc(MQTT_MAX_PACKET_SIZE);
}

CoolPubSubClient::~CoolPubSubClient() {
  free(this->buffer);
}

boolean CoolPubSubClient::connect(const char *id, uint16_t keepAlive, uint16_t socketTimeout) {
    return connect(id,NULL,NULL,0,0,0,0,keepAlive,socketTimeout);
 }

boolean CoolPubSubClient::connect(const char *id, const char *user, const char *pass, uint16_t keepAlive, uint16_t socketTimeout) {
    return connect(id,user,pass,0,0,0,0,keepAlive,socketTimeout);
 }

boolean CoolPubSubClient::connect(const char *id, const char* willTopic, uint8_t willQos, boolean willRetain, const char* willMessage, uint16_t keepAlive, uint16_t socketTimeout) {
    return connect(id,NULL,NULL,willTopic,willQos,willRetain,willMessage,keepAlive,socketTimeout);
 }

boolean CoolPubSubClient::connect(const char *id, const char *user, const char *pass, const char* willTopic, uint8_t willQos, boolean willRetain, const char* willMessage, uint16_t keepAlive, uint16_t socketTimeout) {  
    if (!connected()) {
        int result = 0;

        if (domain != NULL) {
            result = _client->connect(this->domain, this->port);
        } else {
            result = _client->connect(this->ip, this->port);
        }
        if (result == 1) {
            nextMsgId = 1;
            // Leave room in the buffer for header and variable length field
            uint16_t length = 5;
            unsigned int j;

#if MQTT_VERSION == MQTT_VERSION_3_1
            uint8_t d[9] = {0x00,0x06,'M','Q','I','s','d','p', MQTT_VERSION};
#define MQTT_HEADER_VERSION_LENGTH 9
#elif MQTT_VERSION == MQTT_VERSION_3_1_1
            uint8_t d[7] = {0x00,0x04,'M','Q','T','T',MQTT_VERSION};
#define MQTT_HEADER_VERSION_LENGTH 7
#endif
            for (j = 0;j<MQTT_HEADER_VERSION_LENGTH;j++) {
                buffer[length++] = d[j];
            }

            uint8_t v;
            if (willTopic) {
                v = 0x06|(willQos<<3)|(willRetain<<5);
            } else {
                v = 0x02;
            }

            if(user != NULL) {
                v = v|0x80;

                if(pass != NULL) {
                    v = v|(0x80>>1);
                }
            }

            buffer[length++] = v;

            if (keepAlive > 0) {
                this->keepAlive = keepAlive;
            } else {
                this->keepAlive = MQTT_KEEPALIVE;
            }

            buffer[length++] = ((this->keepAlive) >> 8);
            buffer[length++] = ((this->keepAlive) & 0xFF);

            length = writeString(id,buffer,length);
            if (willTopic) {
                length = writeString(willTopic,buffer,length);
                length = writeString(willMessage,buffer,length);
            }

            if(user != NULL) {
                length = writeString(user,buffer,length);
                if(pass != NULL) {
                    length = writeString(pass,buffer,length);
                }
            }

            write(MQTTCONNECT,buffer,length-5);

            lastInActivity = lastOutActivity = millis();
		
	    if (socketTimeout > 0) {
                this->socketTimeout = socketTimeout;
            } else {
                this->socketTimeout = MQTT_SOCKET_TIMEOUT;
            }


            while (!_client->available()) {
                unsigned long t = millis();
                if (t-lastInActivity >= ((int32_t) this->socketTimeout*1000UL)) {
                    _state = MQTT_CONNECTION_TIMEOUT;
                    _client->stop();
                    return false;
                }
            }
            uint8_t llen;
            uint16_t len = readPacket(&llen);

            if (len == 4) {
                if (buffer[3] == 0) {
                    lastInActivity = millis();
                    pingOutstanding = false;
                    _state = MQTT_CONNECTED;
                    return true;
                } else {
                    _state = buffer[3];
                }
            }
            _client->stop();
        } else {
            _state = MQTT_CONNECT_FAILED;
        }
        return false;
    }
    return true;
}

// reads a byte into result
boolean CoolPubSubClient::readByte(uint8_t * result) {
   uint32_t previousMillis = millis();
   while(!_client->available()) 
   {
     uint32_t currentMillis = millis();
     if(currentMillis - previousMillis >= ((int32_t) this->socketTimeout * 1000))
     { 
       return false;
     }
   }
   *result = _client->read();
   return true;
}

// reads a byte into result[*index] and increments index
boolean CoolPubSubClient::readByte(uint8_t * result, uint16_t * index){
  uint16_t current_index = *index;
  uint8_t * write_address = &(result[current_index]);
  if(readByte(write_address)){
    *index = current_index + 1;
    return true;
  }
  return false;
}

uint16_t CoolPubSubClient::readPacket(uint8_t* lengthLength) {
    uint16_t len = 0;
    if(!readByte(buffer, &len)) return 0;
    bool isPublish = (buffer[0]&0xF0) == MQTTPUBLISH;
    uint32_t multiplier = 1;
    uint16_t length = 0;
    uint8_t digit = 0;
    uint16_t skip = 0;
    uint8_t start = 0;

    do {
        if(!readByte(&digit)) return 0;
        buffer[len++] = digit;
        length += (digit & 127) * multiplier;
        multiplier *= 128;
    } while ((digit & 128) != 0);
    *lengthLength = len-1;

    if (isPublish) {
        // Read in topic length to calculate bytes to skip over for Stream writing
        if(!readByte(buffer, &len)) return 0;
        if(!readByte(buffer, &len)) return 0;
        skip = (buffer[*lengthLength+1]<<8)+buffer[*lengthLength+2];
        start = 2;
        if (buffer[0]&MQTTQOS1) {
            // skip message id
            skip += 2;
        }
    }

    for (uint16_t i = start;i<length;i++) {
        if(!readByte(&digit)) return 0;
        if (this->stream) {
            if (isPublish && len-*lengthLength-2>skip) {
                this->stream->write(digit);
            }
        }
        if (len < this->buffer_size) {
            buffer[len] = digit;
        }
        len++;
    }

    if (!this->stream && len > this->buffer_size) {
        len = 0; // This will cause the packet to be ignored.
    }

    return len;
}

boolean CoolPubSubClient::loop() {
    if (connected()) {
        unsigned long t = millis();
        if ((t - lastInActivity > this->keepAlive*1000UL) || (t - lastOutActivity > this->keepAlive*1000UL)) { 
            if (pingOutstanding) {
                this->_state = MQTT_CONNECTION_TIMEOUT;
                _client->stop();
                return false;
            } else {
                buffer[0] = MQTTPINGREQ;
                buffer[1] = 0;
                _client->write(buffer,2);
                lastOutActivity = t;
                lastInActivity = t;
                pingOutstanding = true;
            }
        }
        if (_client->available()) {
            uint8_t llen;
            uint16_t len = readPacket(&llen);
            uint16_t msgId = 0;
            uint8_t *payload;
            if (len > 0) {
                lastInActivity = t;
                uint8_t type = buffer[0]&0xF0;
                if (type == MQTTPUBLISH) {
                    if (callback) {
                        uint16_t tl = (buffer[llen+1]<<8)+buffer[llen+2]; /* topic length in bytes */
                        memmove(buffer+llen+2,buffer+llen+3,tl); /* move topic inside buffer 1 byte to front */
                        buffer[llen+2+tl] = 0; /* end the topic as a 'C' string with \x00 */
                        char *topic = (char*) buffer+llen+2;
                        // msgId only present for QOS>0
                        if ((buffer[0]&0x06) == MQTTQOS1) {
                            msgId = (buffer[llen+3+tl]<<8)+buffer[llen+3+tl+1];
                            payload = buffer+llen+3+tl+2;
                            callback(topic,payload,len-llen-3-tl-2);

                            buffer[0] = MQTTPUBACK;
                            buffer[1] = 2;
                            buffer[2] = (msgId >> 8);
                            buffer[3] = (msgId & 0xFF);
                            _client->write(buffer,4);
                            lastOutActivity = t;

                        } else {
                            payload = buffer+llen+3+tl;
                            callback(topic,payload,len-llen-3-tl);
                        }
                    }
                } else if (type == MQTTPINGREQ) {
                    buffer[0] = MQTTPINGRESP;
                    buffer[1] = 0;
                    _client->write(buffer,2);
                } else if (type == MQTTPINGRESP) {
                    pingOutstanding = false;
                }
            }
        }
        return true;
    }
    return false;
}

boolean CoolPubSubClient::publish(const char* topic, const char* payload) {
    return publish(topic,(const uint8_t*)payload,strlen(payload),false);
}

boolean CoolPubSubClient::publish(const char* topic, const char* payload, boolean retained) {
    return publish(topic,(const uint8_t*)payload,strlen(payload),retained);
}

boolean CoolPubSubClient::publish(const char* topic, const uint8_t* payload, unsigned int plength) {
    return publish(topic, payload, plength, false);
}

boolean CoolPubSubClient::publish(const char* topic, const uint8_t* payload, unsigned int plength, boolean retained) {
    if (connected()) {
        if (this->buffer_size < 5 + 2+strlen(topic) + plength) {
            // Too long
            return false;
        }
        // Leave room in the buffer for header and variable length field
        uint16_t length = 5;
        length = writeString(topic,buffer,length);
        uint16_t i;
        for (i=0;i<plength;i++) {
            buffer[length++] = payload[i];
        }
        uint8_t header = MQTTPUBLISH;
        if (retained) {
            header |= 1;
        }
        return write(header,buffer,length-5);
    }
    return false;
}

boolean CoolPubSubClient::publish_P(const char* topic, const uint8_t* payload, unsigned int plength, boolean retained) {
    uint8_t llen = 0;
    uint8_t digit;
    unsigned int rc = 0;
    uint16_t tlen;
    unsigned int pos = 0;
    unsigned int i;
    uint8_t header;
    unsigned int len;

    if (!connected()) {
        return false;
    }

    tlen = strlen(topic);

    header = MQTTPUBLISH;
    if (retained) {
        header |= 1;
    }
    buffer[pos++] = header;
    len = plength + 2 + tlen;
    do {
        digit = len % 128;
        len = len / 128;
        if (len > 0) {
            digit |= 0x80;
        }
        buffer[pos++] = digit;
        llen++;
    } while(len>0);

    pos = writeString(topic,buffer,pos);

    rc += _client->write(buffer,pos);

    for (i=0;i<plength;i++) {
        rc += _client->write((char)pgm_read_byte_near(payload + i));
    }

    lastOutActivity = millis();

    return rc == tlen + 4 + plength;
}

boolean CoolPubSubClient::write(uint8_t header, uint8_t* buf, uint16_t length) {
    uint8_t lenBuf[4];
    uint8_t llen = 0;
    uint8_t digit;
    uint8_t pos = 0;
    uint16_t rc;
    uint16_t len = length;
    do {
        digit = len % 128;
        len = len / 128;
        if (len > 0) {
            digit |= 0x80;
        }
        lenBuf[pos++] = digit;
        llen++;
    } while(len>0);

    buf[4-llen] = header;
    for (int i=0;i<llen;i++) {
        buf[5-llen+i] = lenBuf[i];
    }

#ifdef MQTT_MAX_TRANSFER_SIZE
    uint8_t* writeBuf = buf+(4-llen);
    uint16_t bytesRemaining = length+1+llen;  //Match the length type
    uint8_t bytesToWrite;
    boolean result = true;
    while((bytesRemaining > 0) && result) {
        bytesToWrite = (bytesRemaining > MQTT_MAX_TRANSFER_SIZE)?MQTT_MAX_TRANSFER_SIZE:bytesRemaining;
        rc = _client->write(writeBuf,bytesToWrite);
        result = (rc == bytesToWrite);
        bytesRemaining -= rc;
        writeBuf += rc;
    }
    return result;
#else
    rc = _client->write(buf+(4-llen),length+1+llen);
    lastOutActivity = millis();
    return (rc == 1+llen+length);
#endif
}

boolean CoolPubSubClient::subscribe(const char* topic) {
    return subscribe(topic, 0);
}

boolean CoolPubSubClient::subscribe(const char* topic, uint8_t qos) {
    if (qos < 0 || qos > 1) {
        return false;
    }
    if (this->buffer_size < 9 + strlen(topic)) {
        // Too long
        return false;
    }
    if (connected()) {
        // Leave room in the buffer for header and variable length field
        uint16_t length = 5;
        nextMsgId++;
        if (nextMsgId == 0) {
            nextMsgId = 1;
        }
        buffer[length++] = (nextMsgId >> 8);
        buffer[length++] = (nextMsgId & 0xFF);
        length = writeString((char*)topic, buffer,length);
        buffer[length++] = qos;
        return write(MQTTSUBSCRIBE|MQTTQOS1,buffer,length-5);
    }
    return false;
}

boolean CoolPubSubClient::unsubscribe(const char* topic) {
    if (this->buffer_size < 9 + strlen(topic)) {
        // Too long
        return false;
    }
    if (connected()) {
        uint16_t length = 5;
        nextMsgId++;
        if (nextMsgId == 0) {
            nextMsgId = 1;
        }
        buffer[length++] = (nextMsgId >> 8);
        buffer[length++] = (nextMsgId & 0xFF);
        length = writeString(topic, buffer,length);
        return write(MQTTUNSUBSCRIBE|MQTTQOS1,buffer,length-5);
    }
    return false;
}

void CoolPubSubClient::disconnect() {
    buffer[0] = MQTTDISCONNECT;
    buffer[1] = 0;
    _client->write(buffer,2);
    _state = MQTT_DISCONNECTED;
    _client->stop();
    lastInActivity = lastOutActivity = millis();
}

uint16_t CoolPubSubClient::writeString(const char* string, uint8_t* buf, uint16_t pos) {
    const char* idp = string;
    uint16_t i = 0;
    pos += 2;
    while (*idp) {
        buf[pos++] = *idp++;
        i++;
    }
    buf[pos-i-2] = (i >> 8);
    buf[pos-i-1] = (i & 0xFF);
    return pos;
}


boolean CoolPubSubClient::connected() {
    boolean rc;
    if (_client == NULL ) {
        rc = false;
    } else {
        rc = (int)_client->connected();
        if (!rc) {
            if (this->_state == MQTT_CONNECTED) {
                this->_state = MQTT_CONNECTION_LOST;
                _client->flush();
                _client->stop();
            }
        }
    }
    return rc;
}

CoolPubSubClient& CoolPubSubClient::setServer(uint8_t * ip, uint16_t port) {
    IPAddress addr(ip[0],ip[1],ip[2],ip[3]);
    return setServer(addr,port);
}

CoolPubSubClient& CoolPubSubClient::setServer(IPAddress ip, uint16_t port) {
    this->ip = ip;
    this->port = port;
    this->domain = NULL;
    return *this;
}

CoolPubSubClient& CoolPubSubClient::setServer(const char * domain, uint16_t port) {
    this->domain = domain;
    this->port = port;
    return *this;
}

CoolPubSubClient& CoolPubSubClient::setCallback(MQTT_CALLBACK_SIGNATURE) {
    this->callback = callback;
    return *this;
}

CoolPubSubClient& CoolPubSubClient::setClient(Client& client){
    this->_client = &client;
    return *this;
}

CoolPubSubClient& CoolPubSubClient::setStream(Stream& stream){
    this->stream = &stream;
    return *this;
}

int CoolPubSubClient::state() {
    return this->_state;
}

boolean CoolPubSubClient::setBufferSize(uint16_t size) {
  this->buffer = (uint8_t*)realloc(this->buffer, size);
  this->buffer_size = size;
  return (this->buffer == NULL);
}

uint16_t CoolPubSubClient::getBufferSize() {
  return this->buffer_size;
}

CoolPubSubClient& CoolPubSubClient::setTimeout(uint16_t socketTimeout){
    this->socketTimeout = socketTimeout;
    return *this;
}

